lib/analytics/course_revisions_csv_builder.rb

```diff
@@ -0,0 +1,103 @@
+# frozen_string_literal: true
+
+require 'csv'
+
+class CourseRevisionsCsvBuilder
+  def initialize(course)
+    @course = course
+    set_revisions
+  end
+
+  def generate_csv
+    csv_data = [CSV_HEADERS]
+    revisions_rows.each do |row|
+      csv_data << row
+    end
+    CSV.generate { |csv| csv_data.each { |line| csv << line } }
+  end
+
+  def revisions_rows
+    @new_revisions.values.map do |revision_data|
+      build_row(revision_data)
+    end
+  end
+
+  def set_revisions
+    @new_revisions = {}
+    @course.all_revisions.includes(:wiki, :article).map do |edit|
+      revision_edits = @new_revisions[edit.article_id] || new_revision(edit)
+      update_title_username(revision_edits, edit)
+      revision_edits[:mw_rev_id] = edit.mw_rev_id
+      revision_edits[:mw_page_id] = edit.mw_page_id
+      revision_edits[:wiki] = edit.wiki.domain
+      update_characters_references_views(revision_edits, edit)
+      revision_edits[:new_article] = true if edit.new_article
+      revision_edits[:deleted] = true if edit.deleted
+      revision_edits[:wp10] = edit.wp10
+      revision_edits[:wp10_previous] = edit.wp10_previous
+      @new_revisions[edit.article_id] = revision_edits
+    end
+  end
+
+  def new_revision(edit)
+    article = edit.article
+    {
+      title: article.title,
+      new_article: false,
+      views: 0,
+      characters: {},
+      references: {},
+      deleted: edit.deleted,
+      wp10: {},
+      wp10_previous: {},
+      article_id: edit.article_id
+    }
+  end
+
+  def update_title_username(revision_edits, edit)
+    revision_edits[:title] = edit.article.title
+    revision_edits[:username] = edit.user.username
```

Comment:
This is making an extra query for every revision, to fetch the user and read the username.

Adding `:user` to the `includes` on line 27 should this N+1 problem.